clippy_lints/src/matches.rs

```diff
@@ -705,25 +705,52 @@ fn check_wild_in_or_pats(cx: &LateContext<'_, '_>, arms: &[Arm<'_>]) {
 }
 
 /// Gets all arms that are unbounded `PatRange`s.
-fn all_ranges<'a, 'tcx>(cx: &LateContext<'a, 'tcx>, arms: &'tcx [Arm<'_>]) -> Vec<SpannedRange<Constant>> {
+fn all_ranges<'a, 'tcx>(
+    cx: &LateContext<'a, 'tcx>,
+    arms: &'tcx [Arm<'_>],
+    ty: Ty<'tcx>,
+) -> Vec<SpannedRange<Constant>> {
     arms.iter()
         .flat_map(|arm| {
             if let Arm {
                 ref pat, guard: None, ..
             } = *arm
             {
-                if let PatKind::Range(ref lhs, ref rhs, ref range_end) = pat.kind {
-                    if let (Some(l), Some(r)) = (lhs, rhs) {
-                        let lhs = constant(cx, cx.tables, l)?.0;
-                        let rhs = constant(cx, cx.tables, r)?.0;
-                        let rhs = match *range_end {
-                            RangeEnd::Included => Bound::Included(rhs),
-                            RangeEnd::Excluded => Bound::Excluded(rhs),
-                        };
-                        return Some(SpannedRange {
-                            span: pat.span,
-                            node: (lhs, rhs),
-                        });
+                if let PatKind::Range(ref lhs, ref rhs, range_end) = pat.kind {
+                    match (lhs, rhs) {
+                        (Some(lhs), Some(rhs)) => {
+                            let lhs = constant(cx, cx.tables, lhs)?.0;
+                            let rhs = constant(cx, cx.tables, rhs)?.0;
+                            let rhs = match range_end {
+                                RangeEnd::Included => Bound::Included(rhs),
+                                RangeEnd::Excluded => Bound::Excluded(rhs),
+                            };
+                            return Some(SpannedRange {
+                                span: pat.span,
+                                node: (lhs, rhs),
+                            });
+                        },
+                        (None, Some(rhs)) => {
+                            let lhs = miri_to_const(ty.numeric_min_val(cx.tcx)?)?;
```

Comment:
A simpler approach could be to use `{l,r}hs.unwrap_or_else(|| ...)` and simply ignore the fact that `(None, None)` is a possibility.

Reply:
Uhm, we can use `map_or` but it requires some `unwrap`s, I think.
like this:

```rust
let lhs = lhs.map_or(
    miri_to_const(ty.numeric_min_val(cx.tcx).unwrap()).unwrap(),
    |l| constant(cx, cx.tables, l).unwrap().0
);
let rhs = rhs.map_or(
    miri_to_const(ty.numeric_min_val(cx.tcx).unwrap()).unwrap(),
    |r| constant(cx, cx.tables, r).unwrap().0
);
```

Is there more appropriate approach?

Reply:
Well you can use `match` on `lhs` or `rhs` individually, and it will still be an improvement.

Reply:
You suggest to collapse `(None, Some(rhs))` and `(Some(lhs), None)`, right?

Reply:
My suggestion is to do:
```rust
let lhs = match lhs {
    Some(expr) => constant(cx, cx.tables, lhs)?.0,
    None => miri_to_const(ty.numeric_min_val(cx.tcx)?)?,
};
```
and similarly with `rhs`, but with `numeric_max_val`.

Reply:
Ah thanks, I've just got your intention, reduced lines via 3bdf40407606b82becacdb1c0f77e73c2f6c9871.